### PR TITLE
CVSL-1724 adjust logic around isInHardStopPeriod

### PR DESCRIPTION
**Pull request by gurps-singh**
This PR is to adjust the logic around isInHardStopPeriod to cater for early releases. This introduces logic to get a hard stop date for a licence as well as adjusting the calculation of whether a licence is in its hard stop period.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt [hunk 1]
```diff
@@ -25,18 +25,15 @@ class ReleaseDateService(
...
       log.warn("Licence with id: ${licence.id} has no CRD or ARD")
       return false
     }
-    return if (releaseDate < LocalDate.now(now)) {
-      false
-    } else {
-      releaseDate <= endOfHardStopPeriod
-    }
+
+    return today >= hardStopDate && today <= licence.licenceStartDate
```
Comment (gurps-singh) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt:36]: Discussed with Seb, using licenceStartDate is doing ARD || CRD and also means that this logic would be false for early releases.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt [hunk 2]
```diff
@@ -60,6 +57,27 @@ class ReleaseDateService(
...
+    val conditionalReleaseDate = licence.conditionalReleaseDate ?: return null
+
+    val hardStopDateUsingCrd = workingDaysService.workingDaysBefore(conditionalReleaseDate).take(maxNumberOfWorkingDaysToUpdateLicenceTimeOutStatus).last()
+
+    if (actualReleaseDate == null) {
+      return hardStopDateUsingCrd
+    }
+
+    val hardStopDateUsingArd = workingDaysService.workingDaysBefore(actualReleaseDate).take(maxNumberOfWorkingDaysToUpdateLicenceTimeOutStatus).last()
+    val isNotAnEarlyRelease = actualReleaseDate >= workingDaysService.workingDaysBefore(conditionalReleaseDate).take(1).last()
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt:71]:
```kt
fun oneWorkingDayBefore(date: LocalDate) 
 = workingDaysService.workingDaysBefore(date).take(1).last()
```
=>
```kt
val isNotAnEarlyRelease = actualReleaseDate >= oneWorkingDayBefore(conditionalReleaseDate)
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt [hunk 3]
```diff
@@ -60,6 +57,27 @@ class ReleaseDateService(
...
+    val isArdTheReleaseDate = actualReleaseDate <= conditionalReleaseDate
+
+    return if (isNotAnEarlyRelease && isArdTheReleaseDate) {
+      hardStopDateUsingArd
+    } else {
+      hardStopDateUsingCrd
+    }
+  }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt:79]: would be tempted to have a 
```kt
val date = chooseDateForHardstop(ard, crd)
return twoWorkingDaysBefore(date)   
```

Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt:82]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt (hunk 2 above)
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt]: Done

**Review by andrewrlee — DISMISSED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt [hunk 4]
```diff
@@ -60,6 +57,34 @@ class ReleaseDateService(
...
+    return if (isNotAnEarlyRelease && isArdTheReleaseDate) {
+      actualReleaseDate
+    } else {
+      conditionalReleaseDate
+    }
+  }
+
+  private fun oneWorkingDayBefore(date: LocalDate) = workingDaysService.workingDaysBefore(date).take(1).last()
+
+  private fun twoWorkingDaysBefore(date: LocalDate) = workingDaysService.workingDaysBefore(date).take(2).last()
```
[the remark this answers is not in the document]
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ReleaseDateService.kt]: I'll be superfly!

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**